feat: add C implementation for `blas/base/sspr`

aman-095 opened the pull request:
Progresses #2039.

## Description

> What is the purpose of this pull request?

This pull request adds C implementation for `sspr`.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #2039 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

aman-095 committed:
feat: add C implementation for sspr

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/README.md:
````diff
@@ -183,21 +183,60 @@ console.log( AP );
 ```c
-TODO
+#include "stdlib/blas/base/sspr.h"
+#include "stdlib/blas/base/shared.h"
````
@aman-095 This include should be removed from this code block and explicitly added to the subsequent code blocks.

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/README.md:
```diff
@@ -183,21 +183,60 @@ console.log( AP );
+-   **uplo**: `[in] CBLAS_UPLO` specifies whether the upper or lower triangular part of the symmetric matrix `A` should be referenced.
+-   **N**: `[in] CBLAS_INT` number of elements along each dimension of `A`.
+-   **alpha**: `[in] float` scalar.
+-   **X**: `[in] float*` first input array.
```
This description is a bit confusing. Should this be `input vector`?

aman-095 replied:
Yes, we have used `input array` elsewhere.

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/README.md:
```diff
@@ -183,21 +183,60 @@ console.log( AP );
+-   **N**: `[in] CBLAS_INT` number of elements along each dimension of `A`.
+-   **alpha**: `[in] float` scalar.
+-   **X**: `[in] float*` first input array.
+-   **strideX**: `[in] CBLAS_INT` index increment for `X`.
```
We should be consistent in our descriptions (e.g., w L235). "stride length for `X`"

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/README.md:
```diff
@@ -183,21 +183,60 @@ console.log( AP );
+-   **X**: `[in] float*` first input array.
+-   **strideX**: `[in] CBLAS_INT` index increment for `X`.
+-   **AP**: `[inout] float*` packed form of a symmetric matrix `A`.
+-   **strideAP**: `[in] CBLAS_INT` `AP` stride length.
```
```suggestion
-   **strideAP**: `[in] CBLAS_INT` stride length for `AP`.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/README.md:
```diff
@@ -219,7 +258,34 @@ TODO
+    float AP[] = { 1.0f, 2.0f, 3.0f, 1.0f, 2.0f, 1.0f };
+    const float x[] = { 1.0f, 2.0f, 3.0f };
+
+    // Specify the number of elements:
```
```suggestion
    // Specify the number of elements along each dimension of `A`:
```

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,180 @@
+	print_version();
+	count = 0;
+	for ( i = MIN; i <= MAX; i++ ) {
+		len = stdlib_base_floorf( pow( pow( 10, i ), 1.0/2.0 ) );
```
```suggestion
		len = floor( pow( pow( 10, i ), 1.0/2.0 ) );
```

Use the built-in `floor` here. No need to bring in an extra dep for this C benchmark.

kgryte replied:
This is especially true given that we are using C99 `pow`.

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/examples/c/example.c:
```diff
@@ -0,0 +1,46 @@
+	float AP[] = { 1.0f, 2.0f, 3.0f, 1.0f, 2.0f, 1.0f };
+	const float x[] = { 1.0f, 2.0f, 3.0f };
+
+	// Specify the number of elements:
```
Same comments as README.

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/lib/ndarray.native.js:
```diff
@@ -0,0 +1,64 @@
+* @param {string} uplo - specifies whether the upper or lower triangular part of the symmetric matrix `A` is supplied
+* @param {NonNegativeInteger} N - number of elements along each dimension of `A`
+* @param {number} alpha - scalar
+* @param {Float32Array} x - first input array
```
Description.

kgryte replied:
This likely applies elsewhere throughout this package.

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/lib/ndarray.native.js:
```diff
@@ -0,0 +1,64 @@
+* @param {NonNegativeInteger} offsetAP - starting index for `AP`
+* @throws {TypeError} first argument must be a valid order
+* @throws {TypeError} second argument must specify whether the lower or upper triangular matrix is supplied
+* @throws {RangeError} third argument must be a nonnegative integer
```
Just checking. Are all these errors actually thrown by the native implementation?

aman-095 replied:
No, they are not. Don't know why I kept these. Updating it.

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/lib/sspr.native.js:
```diff
@@ -0,0 +1,63 @@
+* @param {integer} strideX - `x` stride length
+* @param {Float32Array} AP - packed form of a symmetric matrix `A`
+* @throws {TypeError} first argument must be a valid order
+* @throws {TypeError} second argument must specify whether the lower or upper triangular matrix is supplied
```
Same question.

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/manifest.json:
```diff
@@ -0,0 +1,311 @@
+      ]
+    }
+  ]
+}
```
Trailing newline.

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/src/sspr.c:
```diff
@@ -0,0 +1,108 @@
+* @param strideX  `x` stride length
+* @param AP       packed form of a symmetric matrix `A`
+*/
+void API_SUFFIX(c_sspr)( const CBLAS_LAYOUT order, const CBLAS_UPLO uplo, const CBLAS_INT N, const float alpha, const float *X, const CBLAS_INT strideX, float *AP ) {
```
If `X` is capitalized here, it should also be capitalized in the docs above.

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/src/sspr.c:
```diff
@@ -0,0 +1,108 @@
+		kk += ( N - i1 ) * strideAP;
+	}
+	return;
+}
```
Trailing newline.

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/src/sspr.c:
```diff
@@ -0,0 +1,108 @@
+				ix0 = ox;
+				iap = kk;
+				for ( i0 = 0; i0 <= i1; i0++ ) {
+					AP[ iap ] += ( X[ ix0 ] * tmp );
```
```suggestion
					AP[ iap ] += X[ ix0 ] * tmp;
```

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/src/sspr.c:
```diff
@@ -0,0 +1,108 @@
+			ix0 = ix1;
+			iap = kk;
+			for ( i0 = 0; i0 < N - i1; i0++ ) {
+				AP[ iap ] += ( X[ ix0 ] * tmp );
```
```suggestion
				AP[ iap ] += X[ ix0 ] * tmp;
```

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/src/sspr.c:
```diff
@@ -0,0 +1,108 @@
+void API_SUFFIX(c_sspr)( const CBLAS_LAYOUT order, const CBLAS_UPLO uplo, const CBLAS_INT N, const float alpha, const float *X, const CBLAS_INT strideX, float *AP ) {
+	CBLAS_INT ox;
+
+	if ( N == 0 || alpha == 0.0f ) {
```
Why is this present here but not in the `ndarray` implementation?

aman-095 replied:
I think I missed it. Added.

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/src/sspr.c:
```diff
@@ -0,0 +1,108 @@
+* @param strideAP  stride of the first dimension of `AP`
+* @param offsetAP   starting index of `AP`
+*/
+void API_SUFFIX(c_sspr_ndarray)( const CBLAS_LAYOUT order, const CBLAS_UPLO uplo, const CBLAS_INT N, const float alpha, const float *X, const CBLAS_INT strideX, const CBLAS_INT offsetX, float *AP, const CBLAS_INT strideAP, const CBLAS_INT offsetAP ) {
```
Same comments regarding `X` capitalization in docs. If not desired in docs, lowercase here and in the implementation.

kgryte replied:
The latter may be preferred in order to convey that `x` is a vector. Up to you.

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/src/sspr_cblas.c:
```diff
@@ -0,0 +1,36 @@
+*/
+void API_SUFFIX(c_sspr)( const CBLAS_LAYOUT order, const CBLAS_UPLO uplo, const CBLAS_INT N, const float alpha, const float *X, const CBLAS_INT strideX, float *AP ) {
+	API_SUFFIX(cblas_sspr)( order, uplo, N, alpha, X, strideX, AP );
+}
```
Trailing newline.

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/src/sspr_cblas.c:
```diff
@@ -0,0 +1,36 @@
+* @param strideX  `x` stride length
+* @param AP       packed form of a symmetric matrix `A`
+*/
+void API_SUFFIX(c_sspr)( const CBLAS_LAYOUT order, const CBLAS_UPLO uplo, const CBLAS_INT N, const float alpha, const float *X, const CBLAS_INT strideX, float *AP ) {
```
Same comments.

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/src/sspr_cblas.c:
```diff
@@ -0,0 +1,36 @@
+* @param uplo     specifies whether the upper or lower triangular part of the symmetric matrix `A` should be referenced
+* @param N        number of elements along each dimension of `A`
+* @param alpha    scalar
+* @param x        first input array
```
Same comment regarding desc.

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/test/test.ndarray.native.js:
```diff
@@ -0,0 +1,520 @@
+
+	expected = new Float32Array( data.AP_out );
+
+	out = sspr( str2enumLayout( data.order ), str2enum( data.uplo ), data.N, data.alpha, x, data.strideX, data.offsetX, ap, data.strideAP, data.offsetAP );
```
Why are you providing enums here and below when the implementation takes care of that?

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/test/test.sspr.native.js:
```diff
@@ -0,0 +1,300 @@
+
+	expected = new Float32Array( data.AP_out );
+
+	out = sspr( str2enumLayout( data.order ), str2enum( data.uplo ), data.N, data.alpha, x, data.strideX, ap );
```
Same comment.

kgryte reviewed (changes requested):
Thanks for working on this, @aman-095. Let an initial round of comments.

aman-095 commented:
/stdlib merge

aman-095 commented:
/stdlib update-copyright-years

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into sspr-c

kgryte commented:
/stdlib update-copyright-years

aman-095 committed:
chore: apply review changes

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: missing_dependencies
  - task: lint_c_examples
    status: missing_dependencies
  - task: lint_c_benchmarks
    status: missing_dependencies
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: passed
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: passed
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/README.md:
```diff
@@ -183,21 +183,63 @@ console.log( AP );
+-   **N**: `[in] CBLAS_INT` number of elements along each dimension of `A`.
+-   **alpha**: `[in] float` scalar.
+-   **X**: `[in] float*` input array.
+-   **strideX**: `[in] CBLAS_INT` index increment for `X`.
```
```suggestion
-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/README.md:
```diff
@@ -183,21 +183,63 @@ console.log( AP );
+-   **uplo**: `[in] CBLAS_UPLO` specifies whether the upper or lower triangular part of the symmetric matrix `A` should be referenced.
+-   **N**: `[in] CBLAS_INT` number of elements along each dimension of `A`.
+-   **alpha**: `[in] float` scalar.
+-   **X**: `[in] float*` input array.
```
```suggestion
-   **X**: `[in] float*` input vector.
```

to match JSDoc.

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/README.md:
```diff
@@ -183,21 +183,63 @@ console.log( AP );
+-   **uplo**: `[in] CBLAS_UPLO` specifies whether the upper or lower triangular part of the symmetric matrix `A` should be referenced.
+-   **N**: `[in] CBLAS_INT` number of elements along each dimension of `A`.
+-   **alpha**: `[in] float` scalar.
+-   **X**: `[in] float*` input array.
```
```suggestion
-   **X**: `[in] float*` input vector.
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,179 @@
+	stdlib_strided_sfill( len*(len+1)/2, 1.0f, AP, 1 );
+	t = tic();
+	for ( i = 0; i < iterations; i++ ) {
+		c_sspr( CblasRowMajor, CblasUpper, len, 1.0, x, 1, AP );
```
```suggestion
		c_sspr( CblasRowMajor, CblasUpper, len, 1.0f, x, 1, AP );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,179 @@
+	stdlib_strided_sfill( len*(len+1)/2, 1.0f, AP, 1 );
+	t = tic();
+	for ( i = 0; i < iterations; i++ ) {
+		c_sspr_ndarray( CblasRowMajor, CblasUpper, len, 1.0, x, 1, 0, AP, 1, 0 );
```
```suggestion
		c_sspr_ndarray( CblasRowMajor, CblasUpper, len, 1.0f, x, 1, 0, AP, 1, 0 );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/src/sspr.c:
```diff
@@ -0,0 +1,111 @@
+* @param AP       packed form of a symmetric matrix `A`
+*/
+void API_SUFFIX(c_sspr)( const CBLAS_LAYOUT order, const CBLAS_UPLO uplo, const CBLAS_INT N, const float alpha, const float *X, const CBLAS_INT strideX, float *AP ) {
+	CBLAS_INT ox;
+
+	if ( N == 0 || alpha == 0.0f ) {
+		return;
+	}
+	ox = STDLIB_BLAS_BASE_STRIDE2OFFSET( N, strideX );
```
```suggestion
	const CBLAS_INT ox = STDLIB_BLAS_BASE_STRIDE2OFFSET( N, strideX );
```

kgryte committed:
refactor: remove duplication

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/base/sspr/src/sspr.c:
```diff
@@ -0,0 +1,106 @@
+* @param offsetX   starting index of `x`
+* @param AP        packed form of a symmetric matrix `A`
+* @param strideAP  stride of the first dimension of `AP`
+* @param offsetAP   starting index of `AP`
```
```suggestion
* @param offsetAP  starting index of `AP`
```

kgryte committed:
docs: fix alignment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
